File: hu.bme.mit.theta.core/src/main/java/hu/bme/mit/theta/core/type/abstracttype/SubExpr.java
package hu.bme.mit.theta.core.type.abstracttype;

import hu.bme.mit.theta.core.type.BinaryExpr;
import hu.bme.mit.theta.core.type.Expr;

public abstract class SubExpr<ExprType extends Additive<ExprType>> extends BinaryExpr<ExprType, ExprType> {

	protected SubExpr(final Expr<ExprType> leftOp, final Expr<ExprType> rightOp) {
		super(leftOp, rightOp);
	}

}

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/abstracttype/SubExpr.kt
package hu.bme.mit.theta.core.type.abstracttype

import hu.bme.mit.theta.core.type.BinaryExpr
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.utils.TypeUtils.cast
import kotlinx.serialization.Serializable

/**
 * Abstract base class for subtraction expressions over additive types.
 */
@Serializable
abstract class SubExpr<ExprType : Additive<ExprType>> : BinaryExpr<ExprType, ExprType>() {

    companion object {
        @JvmStatic
        fun <ExprType : Additive<ExprType>> create2(
            leftOp: Expr<*>,
            rightOp: Expr<*>
        ): SubExpr<*> {
            @Suppress("UNCHECKED_CAST")
            val type = leftOp.type as ExprType
            val newLeftOp = cast(leftOp, type)
            val newRightOp = cast(rightOp, type)
            return type.Sub(newLeftOp, newRightOp)
        }
    }
}
